### ISSUE TITLE
scheduled plugin doesn't implement please_stop

### ISSUE THREAD

**Issue by reidsunderland** [MetPX/sarracenia#752]
https://github.com/MetPX/sarracenia/blob/main/sarracenia/flowcb/scheduled/__init__.py

The scheduled plugin should implement please stop. Without it, the flow algorithm will try to stop cleanly but fail because the scheduled plugin just keeps sleeping. Then the process has to be killed.

```
sr3d restart flow/scheduled_hydro_quebec

...

ddsr4: stopping: sending SIGTERM  ( 0 ) Done
ddsr4: Waiting 1 sec. to check if 4 processes stopped (try: 0)
ddsr4: Waiting 2 sec. to check if 4 processes stopped (try: 1)
ddsr4: Waiting 4 sec. to check if 4 processes stopped (try: 2)
ddsr4: Waiting 8 sec. to check if 4 processes stopped (try: 3)
ddsr4: Waiting 16 sec. to check if 4 processes stopped (try: 4)
ddsr4: doing SIGKILL this time
ddsr4: Done
ddsr4: Waiting again...
ddsr4: not responding to SIGKILL:
ddsr4:  19102: ['/usr/bin/python3', '/usr/lib/python3/dist-packages/sarracenia/instance.py', '--no', '1', 'start']
ddsr4:  19103: ['/usr/bin/python3', '/usr/lib/python3/dist-packages/sarracenia/instance.py', '--no', '1', 'start']
ddsr4:  24435: ['/usr/bin/python3', '/usr/lib/python3/dist-packages/sarracenia/instance.py', '--no', '1', 'start']
ddsr4:  24437: ['/usr/bin/python3', '/usr/lib/python3/dist-packages/sarracenia/instance.py', '--no', '1', 'start']
ddsr4: starting: .( 1 ) Done
ddsr4: 
```

I think the comment is also out of sync with the code? https://github.com/MetPX/sarracenia/blob/bcf4f4771f9c379722c3f9623f8341a12675ac4c/sarracenia/flowcb/scheduled/__init__.py#L28-L30

Comma separated hours/minutes doesn't work.

**Comment by reidsunderland** [comment]
```
ddsr4: 2023-08-29 17:55:44,453 [INFO] sarracenia.moth.amqp __putSetup exchange declared: 
ddsr4: 2023-08-29 17:55:44,457 [INFO] sarracenia.flowcb.log __init__ flow initialized with: {'ack', 'gather', 'on_start', 'download', 'destfn', 'please_stop', 'on_stop', 'poll', 'after_post', 'post', 'on_housekeeping', 'metricsReport', 'after_accept', 'send', 'on_report', 'after_work', 'reject', 'do_poll'}
ddsr4: 2023-08-29 17:55:44,458 [ERROR] sarracenia.config check_undeclared_options undeclared option: logrotate
ddsr4: 2023-08-29 17:55:44,458 [INFO] sarracenia.flow run callbacks loaded: ['sarracenia.flowcb.retry.Retry', 'sarracenia.flowcb.housekeeping.resources.Resources', 'post.message', 'scheduled', 'log']
ddsr4: 2023-08-29 17:55:44,458 [INFO] sarracenia.flow run pid: 9466 flow/scheduled_hydro_quebec instance: 1
ddsr4: 2023-08-29 17:55:44,461 [INFO] sarracenia.flowcb.log stats version: 3.00.37, started: now, last_housekeeping:  0.0 seconds ago
ddsr4: 2023-08-29 17:55:44,461 [INFO] sarracenia.flowcb.log stats messages received: 0, accepted: 0, rejected: 0   rate accepted: 0.0% or 0.0 m/s
ddsr4: 2023-08-29 17:55:44,461 [INFO] sarracenia.flowcb.log stats files transferred: 0 bytes: 0 Bytes rate: 0 Bytes/sec
ddsr4: 2023-08-29 17:55:44,461 [INFO] sarracenia.flowcb.log on_start starting
ddsr4: 2023-08-29 17:55:44,464 [INFO] sarracenia.flowcb.scheduled wait_until_hour sleep for 68400 sec
```

something else is wrong too, starting this config at 17:55 shouldn't cause it to sleep for 68400 seconds (19 hours). It should only sleep until 18:45...


```
# broker removed

# needed for flow config to post messages
callback post.message

# use a scheduled post of the same files every time
callback scheduled

# files are updated at 12:45, 15:45, 18:45 and 21:45 UTC each day
scheduled_hour   12
scheduled_hour   15
scheduled_hour   18
scheduled_hour   21
scheduled_minute 45

# logging config
logEvents      all
logMessageDump on
callback       log
logLevel       debug

# data source
post_baseUrl https://www.hydroquebec.com/data/documents-donnees/donnees-ouvertes/json/
path Donnees_VUE_STATIONS_ET_TARAGES.json
path Donnees_VUE_CENTRALES_ET_OUVRAGES.json

# no filtering applied to source data
acceptUnmatched on
```

**Comment by petersilva** [comment]
## stop/abort processing

In the source given above, the long waits are a collection of short waits (wait_until_next, wait_until_hour, wait_within_hour) are all implemented as calls that eventually lead to wait_seconds() which only naps for 10 seconds before looking at self.stop_requested again.  

*def please_stop(self):* is implemented in the parent flowcb/__init__.py module/class. it sets stop_requested to true, which should then be picked up by scheduled... so that _should_ work, I remember testing that kind of thing, but I may have missed something.

you should be able to check by seeing the *registering* messages at startup with --debug, it will list each entry point loaded for each callback module.  It should list *please_stop* as a registered entry_point for flowcb/scheduled.

## scheduling

I invoked the flowcb/scheduled/__init__.py from the command line on a linux machine, and it did what seemed sensible.
but it doesn't have the signal set as per when called from sr3.  it was doing minutes within the current hour, then I tried
your config, and it wanted to wait -18000 ... oh dear... I will take a look.

**Comment by reidsunderland** [comment]
I overlooked the stop_requested variable. I guess the reason it isn't stopping is just because it's never getting to check that variable's state because it's sleeping so long. 

I thought the plugin itself had to implement a please_stop method, but the FlowCB parent has it. 

Thank you for looking into it, this isn't urgent. I set the config to use scheduled_interval for now, so it's still getting the data. :)

**Comment by petersilva** [comment]
sigh... many issues...:
* big stupid mistake... in wiat_until_hour ... comparing now.tm_min instead of now.tm_hour... fairly random stuff results.
* smaller stupid mistake >= hour... should be >.
* comma stuff in the config file seems to work fine, I don't see any issue using your config with scheduled_hour 12,15,18,21  and it seems to be parsed fine.
* another timing calculation error: it was always waiting full hours. now it waits until the top of the hour in question, which is necessary for the wait_within_hour() to work properly.
* another timing calculation error: it was always waiting full minutes. Now it waits until the top of the the minute for the next interval.
* still finding stuff.

**Comment by petersilva** [comment]
The second patch (replaces the first, a different branch does not the first one)  Algorithm completely different leverages datetime to build a list of appointments for the day, and then waiting until each appointment in the list.

It should be easy to build unit tests for it, because you can instantiate the plugin, with different settings, and set a time of day and call update_appointments( when ) and the list of appointments should match the settings (should be able to test many settings without waiting for any of them.

I noticed that housekeeping calls don't happen when waiting for an event, so will work on that...

**Comment by petersilva** [comment]
## Housekeeping 
support for interrupting sleeps to do housekeeping added.

**Comment by petersilva** [comment]
released in 3.0.44
